frameworks.qmd

```diff
@@ -1,119 +1,1785 @@
+
+### Optimization Algorithms
+
+Training a neural network is fundamentally an iterative process that
+seeks to minimize a loss function. At its core, the goal is to fine-tune
+the model weights and parameters to produce predictions as close as
+possible to the true target labels. Machine learning frameworks have
+greatly streamlined this process by offering extensive support in three
+critical areas: loss functions, optimization algorithms, and
+regularization techniques.
+
+Loss Functions are useful to quantify the difference between the
+model\'s predictions and the true values. Different datasets require a
+different loss function to perform properly, as the loss function tells
+the computer the "objective" for it to aim to. Commonly used loss
+functions are Mean Squared Error (MSE) for regression tasks and
+Cross-Entropy Loss for classification tasks.
+
+Once the loss is computed, we need methods to adjust the model\'s
+parameters to reduce this loss or error during the training process. To
+do so, current frameworks use a gradient based approach, where it
+computes how much changes tuning the weights in a certain way changes
+the value of the loss function. Knowing this gradient, the model moves
+in the direction that reduces the gradient. There are many challenges
+associated with this, however, primarily stemming from the fact that the
+optimization problem is not convex, making it very easy to solve, and
+more details about this will come in the AI training sections. Modern
+frameworks come equipped with efficient implementations of several
+optimization algorithms, many of which are variants of gradient descent
+algorithms with stochastic methods and adaptive learning rates. More
+information can be found in the AI training section.
+
+Last but not least, overly complex models tend to overfit, meaning they
+perform well on the training data but fail to generalize to new, unseen
+data (see Overfitting). To counteract this, regularization methods are
+employed to penalize model complexity and encourage it to learn simpler
+patterns. Dropout for instance randomly sets a fraction of input units
+to 0 at each update during training, which helps prevent overfitting.
+
+Frameworks also provide efficient implementations of gradient descent,
+Adagrad, Adadelta, and Adam. Adding regularization like dropout and
+L1/L2 penalties prevents overfitting during training. Batch
+normalization accelerates training by normalizing inputs to layers.
+
+### Model Training Support
+
+Before training a defined neural network model, a compilation step is
+required. During this step, the high-level architecture of the neural
+network is transformed into an optimized, executable format. This
+process comprises several steps. The construction of the computational
+graph is the first step. It represents all the mathematical operations
+and data flow within the model. We discussed this earlier.
+
+During training, the focus is on executing the computational graph.
```

Comment:
Computational graph might be familiar to people who are exposed to AI/ML, but it doesn't mean much to people who aren't experienced in the field. Since this textbook covers this topic, I think a graphical representation of a computational graph would be good to include here!

Reply:
Discussed this earlier. Thanks!